File: src/lib/engine/tuning.ts
/**
 * De frecuencia a nota y viceversa: lo que necesita un afinador.
 *
 * Todo depende de una sola referencia (La4, 440 Hz por defecto). Cambiarla
 * mueve las doce notas a la vez, que es exactamente lo que hace falta para
 * tocar con alguien afinado en 432 o con un piano viejo.
 */

import { Midi, PC_NAMES_SHARP, Tuning, midiToPc } from "./notes";

export const DEFAULT_A4 = 440;

export function midiToFrequency(midi: Midi, a4 = DEFAULT_A4): number {
  return a4 * Math.pow(2, (midi - 69) / 12);
}

/** Nota MIDI con decimales: 60.5 es medio semitono arriba de C4. */
export function frequencyToMidiFloat(frequency: number, a4 = DEFAULT_A4): number {
  return 69 + 12 * Math.log2(frequency / a4);
}

/** Diferencia en cents entre dos frecuencias (positivo = la primera es más aguda). */
export function centsBetween(frequency: number, reference: number): number {
  return 1200 * Math.log2(frequency / reference);
}

export interface NoteReading {
  frequency: number;
  /** Nota temperada más cercana. */
  midi: Midi;
  /** "A", "C#"… */
  noteName: string;
  octave: number;
  /** "A3", "C#4". */
  fullName: string;
  /** Desvío respecto de esa nota, entre −50 y +50. */
  cents: number;
}

/** Qué nota es esta frecuencia y cuánto le falta o le sobra. */
export function readNote(frequency: number, a4 = DEFAULT_A4): NoteReading {
  const exact = frequencyToMidiFloat(frequency, a4);
  const midi = Math.round(exact);
  return {
    frequency,
    midi,
    noteName: PC_NAMES_SHARP[midiToPc(midi)],
    octave: Math.floor(midi / 12) - 1,
    fullName: `${PC_NAMES_SHARP[midiToPc(midi)]}${Math.floor(midi / 12) - 1}`,
    cents: (exact - midi) * 100,
  };
}

export interface StringTarget {
  /** Índice de cuerda, 0 = la más grave. */
  index: number;
  midi: Midi;
  /** "D", "G", "B", "E". */
  label: string;
  /** "D3", "G3"… */
  fullName: string;
  frequency: number;
}

/** Las cuerdas al aire de una afinación, con su frecuencia objetivo. */
export function stringTargets(tuning: Tuning, a4 = DEFAULT_A4): StringTarget[] {
  return tuning.strings.map((midi, index) => ({
    index,
    midi,
    label: tuning.labels[index],
    fullName: `${PC_NAMES_SHARP[midiToPc(midi)]}${Math.floor(midi / 12) - 1}`,
    frequency: midiToFrequency(midi, a4),
  }));
}

export interface StringMatch {
  target: StringTarget;
  /** Desvío respecto de esa cuerda (puede pasar de ±50 si está muy floja). */
  cents: number;
}

/**
 * Qué cuerda estás tocando.
 *
 * Se elige la más cercana en cents, pero solo si está dentro de `maxCents`:
 * una cuerda a más de tres semitonos de su nota no es "esa cuerda desafinada",
 * es otra nota, y decir lo contrario haría girar la clavija para el lado
 * equivocado.
 */
export function nearestString(
  frequency: number,
  targets: StringTarget[],
  maxCents = 350,
): StringMatch | null {
  let best: StringMatch | null = null;
  for (const target of targets) {
    const cents = centsBetween(frequency, target.frequency);
    if (!best || Math.abs(cents) < Math.abs(best.cents)) best = { target, cents };
  }
  if (!best || Math.abs(best.cents) > maxCents) return null;
  return best;
}

export type TuningVerdict = "afinada" | "baja" | "alta";

/** Tolerancia estándar de un afinador: ±5 cents no se distingue de oído. */
export const IN_TUNE_CENTS = 5;

export function verdictFor(cents: number, tolerance = IN_TUNE_CENTS): TuningVerdict {
  if (Math.abs(cents) <= tolerance) return "afinada";
  return cents < 0 ? "baja" : "alta";
}

/**
 * Qué hacer con la clavija.
 *
 * Cuando el desvío se sale del dial, el indicador queda pegado al borde y un
 * semitono entero parecería un error chiquito: por eso el texto avisa que hay
 * que girar bastante, no un toque.
 */
export function instructionFor(verdict: TuningVerdict, cents = 0): string {
  const far = Math.abs(cents) > 50;
  switch (verdict) {
    case "afinada":
      return "Afinada";
    case "baja":
      return far ? "Tensá bastante la cuerda" : "Tensá la cuerda";
    case "alta":
      return far ? "Aflojá bastante la cuerda" : "Aflojá la cuerda";
  }
}

File: mobile/targets/watch/Tuning.swift
import Foundation

/// De frecuencia a nota y viceversa. Traducción de `src/lib/engine/tuning.ts`
/// y de lo mínimo de `notes.ts` que el afinador necesita.

let defaultA4: Double = 440
/// Tolerancia estándar de un afinador: ±5 cents no se distingue de oído.
let inTuneCents: Double = 5

private let pcNamesSharp = ["C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"]

func midiToFrequency(_ midi: Int, a4: Double = defaultA4) -> Double {
    a4 * pow(2, Double(midi - 69) / 12)
}

func frequencyToMidiFloat(_ frequency: Double, a4: Double = defaultA4) -> Double {
    69 + 12 * log2(frequency / a4)
}

/// Diferencia en cents entre dos frecuencias (positivo = la primera es más aguda).
func centsBetween(_ frequency: Double, _ reference: Double) -> Double {
    1200 * log2(frequency / reference)
}

struct NoteReading {
    let frequency: Double
    let midi: Int
    let noteName: String
    let octave: Int
    /// "A3", "C#4".
    let fullName: String
    /// Desvío respecto de esa nota, entre −50 y +50.
    let cents: Double
}

func readNote(_ frequency: Double, a4: Double = defaultA4) -> NoteReading {
    let exact = frequencyToMidiFloat(frequency, a4: a4)
    let midi = Int(exact.rounded())
    let pc = ((midi % 12) + 12) % 12
    let octave = Int(floor(Double(midi) / 12)) - 1
    return NoteReading(
        frequency: frequency,
        midi: midi,
        noteName: pcNamesSharp[pc],
        octave: octave,
        fullName: "\(pcNamesSharp[pc])\(octave)",
        cents: (exact - Double(midi)) * 100
    )
}

struct StringTarget: Identifiable, Equatable {
    /// Índice de cuerda, 0 = la más grave.
    let index: Int
    let midi: Int
    /// "D", "G", "B", "E".
    let label: String
    /// "D3", "G3"…
    let fullName: String
    let frequency: Double
    var id: Int { index }
}

/// Ukelele barítono estándar: D3–G3–B3–E4 (grave → agudo).
let baritoneStrings: [(midi: Int, label: String)] = [(50, "D"), (55, "G"), (59, "B"), (64, "E")]

func stringTargets(a4: Double = defaultA4) -> [StringTarget] {
    baritoneStrings.enumerated().map { index, s in
        let pc = ((s.midi % 12) + 12) % 12
        let octave = Int(floor(Double(s.midi) / 12)) - 1
        return StringTarget(
            index: index,
            midi: s.midi,
            label: s.label,
            fullName: "\(pcNamesSharp[pc])\(octave)",
            frequency: midiToFrequency(s.midi, a4: a4)
        )
    }
}

struct StringMatch {
    let target: StringTarget
    let cents: Double
}

/// Qué cuerda estás tocando: la más cercana en cents, solo si está dentro de
/// `maxCents`. Más lejos no es "esa cuerda desafinada", es otra nota.
func nearestString(_ frequency: Double, targets: [StringTarget], maxCents: Double = 350) -> StringMatch? {
    var best: StringMatch?
    for target in targets {
        let cents = centsBetween(frequency, target.frequency)
        if best == nil || abs(cents) < abs(best!.cents) { best = StringMatch(target: target, cents: cents) }
    }
    guard let match = best, abs(match.cents) <= maxCents else { return nil }
    return match
}

enum TuningVerdict {
    case inTune, low, high
}

func verdictFor(_ cents: Double, tolerance: Double = inTuneCents) -> TuningVerdict {
    if abs(cents) <= tolerance { return .inTune }
    return cents < 0 ? .low : .high
}

/// Qué hacer con la clavija. Si el desvío se sale del dial, avisa que hay que
/// girar bastante, no un toque.
func instructionFor(_ verdict: TuningVerdict, cents: Double = 0) -> String {
    let far = abs(cents) > 50
    switch verdict {
    case .inTune: return "Afinada"
    case .low: return far ? "Tensá bastante" : "Tensá la cuerda"
    case .high: return far ? "Aflojá bastante" : "Aflojá la cuerda"
    }
}
